omniledger/service/collect_tx.go

```diff
@@ -0,0 +1,119 @@
+package service
+
+import (
+	"errors"
+	"time"
+
+	"github.com/dedis/cothority/skipchain"
+	"github.com/dedis/onet"
+	"github.com/dedis/onet/network"
+)
+
+func init() {
+	network.RegisterMessages(CollectTxRequest{}, CollectTxResponse{})
+}
+
+// CollectTxProtocol is a protocol for collecting pending transactions.
+type CollectTxProtocol struct {
+	*onet.TreeNodeInstance
+	TxsChan      chan ClientTransactions
+	SkipchainID  skipchain.SkipBlockID
+	requestChan  chan structCollectTxRequest
+	responseChan chan structCollectTxResponse
+	getTxs       func(*network.ServerIdentity, skipchain.SkipBlockID) ClientTransactions
+}
+
+// CollectTxRequest is the request message that asks the receiver to send their
+// pending transactions back to the leader.
+type CollectTxRequest struct {
+	SkipchainID skipchain.SkipBlockID
+}
+
+// CollectTxResponse is the response message that contains all the pending
+// transactions on the node.
+type CollectTxResponse struct {
+	Txs ClientTransactions
+}
+
+type structCollectTxRequest struct {
+	*onet.TreeNode
+	CollectTxRequest
+}
+
+type structCollectTxResponse struct {
+	*onet.TreeNode
+	CollectTxResponse
+}
+
+// NewCollectTxProtocol is used for registering the protocol.
+func NewCollectTxProtocol(getTxs func(*network.ServerIdentity, skipchain.SkipBlockID) ClientTransactions) func(*onet.TreeNodeInstance) (onet.ProtocolInstance, error) {
+	return func(node *onet.TreeNodeInstance) (onet.ProtocolInstance, error) {
+		c := &CollectTxProtocol{
+			TreeNodeInstance: node,
+			TxsChan:          make(chan ClientTransactions),
+			getTxs:           getTxs,
+		}
+		if err := node.RegisterChannels(&c.requestChan, &c.responseChan); err != nil {
+			return c, err
+		}
+		return c, nil
+	}
+}
+
+// Start starts the protocol, it should only be called on the root node.
+func (p *CollectTxProtocol) Start() error {
+	if !p.IsRoot() {
+		return errors.New("only the root should call start")
+	}
+	if len(p.SkipchainID) == 0 {
+		return errors.New("missing skipchain ID")
+	}
+	req := &CollectTxRequest{
+		SkipchainID: p.SkipchainID,
+	}
+	// send to myself and the children
+	if err := p.SendTo(p.TreeNode(), req); err != nil {
+		return err
+	}
+	return p.SendToChildren(req)
+}
+
+// Dispatch runs the protocol.
+func (p *CollectTxProtocol) Dispatch() error {
+	defer p.Done()
+
+	var req structCollectTxRequest
+	select {
+	case req = <-p.requestChan:
+	case <-time.After(time.Second):
+		// This timeout checks whether the root started the protocol,
+		// it is not like our usual timeout that detect failures.
+		return errors.New("did not receive request")
+	}
+
+	// send the result of the callback to the root
+	resp := &CollectTxResponse{
+		Txs: p.getTxs(req.ServerIdentity, req.SkipchainID),
+	}
+	if p.IsRoot() {
+		if err := p.SendTo(p.TreeNode(), resp); err != nil {
+			return err
```

Comment:
If either of these error cases happen, then the transactions which were taken get lost, because they are not re-put back.

Reply:
This is the intended behaviour when I was implementing it. We cannot guarantee every transaction that gets submitted to OL will eventually be included in the skipchain. It'll be even harder to have this guarantee when the leader starts to change (I don't know how to do it yet). But in this particular case, it should be possible to put the transactions back if there is an error.

Reply:
Created an issue to track this problem https://github.com/dedis/cothority/issues/1337